File: Python/14-Backtracking/Algorithms/2-SudokuSolver/SudokuSolver.py
import time
from copy import deepcopy

# Sample Sudoku board (0 represents empty cells)
sample_board = [
    [5, 3, 0, 0, 7, 0, 0, 0, 0],
    [6, 0, 0, 1, 9, 5, 0, 0, 0],
    [0, 9, 8, 0, 0, 0, 0, 6, 0],
    [8, 0, 0, 0, 6, 0, 0, 0, 3],
    [4, 0, 0, 8, 0, 3, 0, 0, 1],
    [7, 0, 0, 0, 2, 0, 0, 0, 6],
    [0, 6, 0, 0, 0, 0, 2, 8, 0],
    [0, 0, 0, 4, 1, 9, 0, 0, 5],
    [0, 0, 0, 0, 8, 0, 0, 7, 9]
]

def print_board(board):
    """Helper function to print the Sudoku board"""
    for i in range(len(board)):
        if i % 3 == 0 and i != 0:
            print("- - - - - - - - - - -")
        for j in range(len(board[0])):
            if j % 3 == 0 and j != 0:
                print("|", end=" ")
            print(board[i][j] if board[i][j] != 0 else ".", end=" ")
        print()

def find_empty(board):
    """Find the next empty cell in the board (row, col)"""
    for i in range(9):
        for j in range(9):
            if board[i][j] == 0:
                return (i, j)
    return None

# Approach 1: Basic Backtracking
def solve_sudoku_basic(board):
    """Basic backtracking approach without any optimizations"""
    empty = find_empty(board)
    if not empty:
        return True  # Puzzle solved

    row, col = empty

    for num in range(1, 10):
        if is_valid(board, num, (row, col)):
            board[row][col] = num

            if solve_sudoku_basic(board):
                return True

            board[row][col] = 0  # Backtrack

    return False

# Approach 2: Backtracking with MRV (Minimum Remaining Values)
def solve_sudoku_mrv(board):
    """Backtracking with Minimum Remaining Values heuristic"""
    empty = find_empty_mrv(board)
    if not empty:
        return True

    row, col = empty

    for num in get_possible_numbers(board, row, col):
        board[row][col] = num

        if solve_sudoku_mrv(board):
            return True

        board[row][col] = 0

    return False

def find_empty_mrv(board):
    """Find empty cell with minimum remaining values (MRV)"""
    min_options = 10
    best_cell = None

    for i in range(9):
        for j in range(9):
            if board[i][j] == 0:
                options = len(get_possible_numbers(board, i, j))
                if options < min_options:
                    min_options = options
                    best_cell = (i, j)

    return best_cell

def get_possible_numbers(board, row, col):
    """Get possible numbers for a cell based on current board state"""
    used_numbers = set()

    # Check row and column
    for i in range(9):
        used_numbers.add(board[row][i])
        used_numbers.add(board[i][col])

    # Check 3x3 box
    box_row = (row // 3) * 3
    box_col = (col // 3) * 3
    for i in range(3):
        for j in range(3):
            used_numbers.add(board[box_row + i][box_col + j])

    return [num for num in range(1, 10) if num not in used_numbers]

# Approach 3: Backtracking with Forward Checking
def solve_sudoku_forward(board):
    """Backtracking with forward checking"""
    empty = find_empty(board)
    if not empty:
        return True

    row, col = empty

    for num in range(1, 10):
        if is_valid(board, num, (row, col)):
            board[row][col] = num

            # Forward checking: remove num from neighbors' domains
            if forward_check(board, row, col, num):
                if solve_sudoku_forward(board):
                    return True

            board[row][col] = 0  # Backtrack

    return False

def forward_check(board, row, col, num):
    """Check if assigning num to (row,col) leaves no empty domains"""
    # Check row
    for j in range(9):
        if board[row][j] == 0 and j != col:
            if not get_possible_numbers(board, row, j):
                return False

    # Check column
    for i in range(9):
        if board[i][col] == 0 and i != row:
            if not get_possible_numbers(board, i, col):
                return False

    # Check 3x3 box
    box_row = (row // 3) * 3
    box_col = (col // 3) * 3
    for i in range(3):
        for j in range(3):
            if board[box_row + i][box_col + j] == 0 and not (box_row + i == row and box_col + j == col):
                if not get_possible_numbers(board, box_row + i, box_col + j):
                    return False

    return True

# Approach 4: Backtracking with Arc Consistency (AC-3)
def solve_sudoku_ac3(board):
    """Backtracking with Arc Consistency (AC-3) algorithm"""
    # Initialize domains
    domains = {}
    for i in range(9):
        for j in range(9):
            if board[i][j] == 0:
                domains[(i, j)] = set(range(1, 10))
            else:
                domains[(i, j)] = {board[i][j]}

    # Run AC-3 to reduce domains
    if not ac3(board, domains):
        return False

    return backtrack_ac3(board, domains)

def backtrack_ac3(board, domains):
    """Backtracking with reduced domains from AC-3"""
    empty = find_empty(board)
    if not empty:
        return True

    row, col = empty

    for num in list(domains[(row, col)]):
        board[row][col] = num
        old_domains = deepcopy(domains)
        domains[(row, col)] = {num}

        # Propagate constraints
        if ac3(board, domains) and backtrack_ac3(board, domains):
            return True

        # Restore domains if failed
        board[row][col] = 0
        domains = old_domains

    return False

def ac3(board, domains):
    """AC-3 algorithm for constraint propagation"""
    queue = []

    # Initialize queue with all constraints
    for i in range(9):
        for j in range(9):
            if board[i][j] == 0:
                # Add row constraints
                for k in range(9):
                    if k != j and board[i][k] == 0:
                        queue.append(((i, j), (i, k)))
                # Add column constraints
                for k in range(9):
                    if k != i and board[k][j] == 0:
                        queue.append(((i, j), (k, j)))
                # Add box constraints
                box_row = (i // 3) * 3
                box_col = (j // 3) * 3
                for x in range(3):
                    for y in range(3):
                        cell = (box_row + x, box_col + y)
                        if cell != (i, j) and board[cell[0]][cell[1]] == 0:
                            queue.append(((i, j), cell))

    while queue:
        (xi, xj), (yi, yj) = queue.pop(0)
        if revise(domains, (xi, xj), (yi, yj)):
            if not domains[(xi, xj)]:
                return False
            # Add new constraints to queue
            # Row neighbors
            for k in range(9):
                if k != xj and board[xi][k] == 0 and k != yj:
                    queue.append(((xi, k), (xi, xj)))
            # Column neighbors
            for k in range(9):
                if k != xi and board[k][xj] == 0 and k != yi:
                    queue.append(((k, xj), (xi, xj)))
            # Box neighbors
            box_row = (xi // 3) * 3
            box_col = (xj // 3) * 3
            for x in range(3):
                for y in range(3):
                    cell = (box_row + x, box_col + y)
                    if cell != (xi, xj) and board[cell[0]][cell[1]] == 0 and cell != (yi, yj):
                        queue.append((cell, (xi, xj)))

    return True

def revise(domains, xi, xj):
    """Revise domains based on constraints"""
    revised = False
    for x in list(domains[xi]):
        # If no value in xj's domain satisfies the constraint (x != y)
        if all(x == y for y in domains[xj]):
            domains[xi].remove(x)
            revised = True
    return revised

# Approach 5: Backtracking with Naked Singles
def solve_sudoku_naked_singles(board):
    """Backtracking that first fills in naked singles (cells with only one possible value)"""
    while True:
        naked_single_found = False
        for i in range(9):
            for j in range(9):
                if board[i][j] == 0:
                    possible = get_possible_numbers(board, i, j)
                    if len(possible) == 1:
                        board[i][j] = possible[0]
                        naked_single_found = True

        if not naked_single_found:
            break

    # Now proceed with regular backtracking
    empty = find_empty(board)
    if not empty:
        return True

    row, col = empty

    for num in get_possible_numbers(board, row, col):
        board[row][col] = num

        if solve_sudoku_naked_singles(board):
            return True

        board[row][col] = 0

    return False

# Common helper function
def is_valid(board, num, pos):
    """Check if placing num at pos is valid"""
    row, col = pos

    # Check row
    for j in range(9):
        if board[row][j] == num and col != j:
            return False

    # Check column
    for i in range(9):
        if board[i][col] == num and row != i:
            return False

    # Check 3x3 box
    box_row = (row // 3) * 3
    box_col = (col // 3) * 3
    for i in range(3):
        for j in range(3):
            if board[box_row + i][box_col + j] == num and (box_row + i != row or box_col + j != col):
                return False

    return True

# Test all approaches
def test_approach(approach_name, solver_func):
    print(f"\nSolving with {approach_name}...")
    board = deepcopy(sample_board)
    print("\nInitial board:")
    print_board(board)

    start_time = time.time()
    if solver_func(board):
        print("\nSolved board:")
        print_board(board)
    else:
        print("\nNo solution exists!")

    elapsed = time.time() - start_time
    print(f"Time taken: {elapsed:.4f} seconds")

if __name__ == "__main__":
    # Test all approaches
    test_approach("Basic Backtracking", solve_sudoku_basic)
    test_approach("Backtracking with MRV", solve_sudoku_mrv)
    test_approach("Backtracking with Forward Checking", solve_sudoku_forward)
    test_approach("Backtracking with AC-3", solve_sudoku_ac3)
    test_approach("Backtracking with Naked Singles", solve_sudoku_naked_singles)

File: C++/14-Backtracking/Algorithms/2-SudokuSolver/SudokuSolver.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <set>
#include <chrono> // For timing
#include <queue> // Required for AC-3

using namespace std;

// Function to print the Sudoku board
void printBoard(const vector<vector<int>>& board) {
    for (int i = 0; i < 9; ++i) {
        if (i % 3 == 0 && i != 0) {
            cout << "--------+---------+--------" << endl;
        }
        for (int j = 0; j < 9; ++j) {
            if (j % 3 == 0 && j != 0) {
                cout << "| ";
            }
            cout << board[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;
}

// Function to check if a number can be placed in a cell
bool isValid(const vector<vector<int>>& board, int row, int col, int num) {
    // Check row
    for (int x = 0; x < 9; ++x) {
        if (board[row][x] == num) {
            return false;
        }
    }

    // Check column
    for (int x = 0; x < 9; ++x) {
        if (board[x][col] == num) {
            return false;
        }
    }

    // Check 3x3 box
    int startRow = row - row % 3;
    int startCol = col - col % 3;
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            if (board[startRow + i][startCol + j] == num) {
                return false;
            }
        }
    }

    return true;
}

// 1. Backtracking (Basic)
bool solveSudokuBacktracking(vector<vector<int>>& board) {
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                for (int num = 1; num <= 9; ++num) {
                    if (isValid(board, row, col, num)) {
                        board[row][col] = num;

                        if (solveSudokuBacktracking(board)) {
                            return true;
                        }
                        else {
                            board[row][col] = 0; // Backtrack
                        }
                    }
                }
                return false; // No valid number found
            }
        }
    }
    return true; // Puzzle solved
}

// 2. Backtracking with Forward Checking
bool solveSudokuForwardCheckingHelper(vector<vector<int>>& board, vector<vector<set<int>>>& possibleValues) {
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                // Iterate through possible values for the current cell.
                for (int num : possibleValues[row][col]) {
                    if (isValid(board, row, col, num)) {
                        board[row][col] = num;

                        // Create a copy to prevent modifying the original
                        vector<vector<set<int>>> nextPossibleValues = possibleValues;
                        // Remove 'num' from possible values of cells in the same row, column, and block
                        for (int i = 0; i < 9; ++i) {
                            nextPossibleValues[row][i].erase(num);
                            nextPossibleValues[i][col].erase(num);
                        }
                        int startRow = row - row % 3;
                        int startCol = col - col % 3;
                        for (int i = 0; i < 3; ++i) {
                            for (int j = 0; j < 3; ++j) {
                                nextPossibleValues[startRow + i][startCol + j].erase(num);
                            }
                        }
                        if (solveSudokuForwardCheckingHelper(board, nextPossibleValues)) {
                            return true;
                        }
                        else {
                            board[row][col] = 0; // Backtrack
                        }
                    }
                }
                return false;
            }
        }
    }
    return true;
}

bool solveSudokuForwardChecking(vector<vector<int>>& board) {
    // possibleValues[row][col] stores the possible numbers that can be placed at board[row][col]
    vector<vector<set<int>>> possibleValues(9, vector<set<int>>(9));

    // Initialize possible values for each cell
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                for (int num = 1; num <= 9; ++num) {
                    if (isValid(board, row, col, num)) {
                        possibleValues[row][col].insert(num);
                    }
                }
            }
        }
    }

    return solveSudokuForwardCheckingHelper(board, possibleValues);
}

// 3. Backtracking with Minimum Remaining Values (MRV) Heuristic
bool solveSudokuMRV(vector<vector<int>>& board) {
    int minRow = -1, minCol = -1, minRemaining = 10; // Initialize minRemaining to a value > 9

    // Find the cell with the minimum remaining possible values
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                int count = 0;
                for (int num = 1; num <= 9; ++num) {
                    if (isValid(board, row, col, num)) {
                        count++;
                    }
                }
                if (count < minRemaining) {
                    minRemaining = count;
                    minRow = row;
                    minCol = col;
                }
            }
        }
    }

    // If no empty cell is found, the puzzle is solved
    if (minRow == -1 && minCol == -1) {
        return true;
    }

    for (int num = 1; num <= 9; ++num) {
        if (isValid(board, minRow, minCol, num)) {
            board[minRow][minCol] = num;
            if (solveSudokuMRV(board)) {
                return true;
            }
            else {
                board[minRow][minCol] = 0;
            }
        }
    }
    return false;
}

// 4. Backtracking with Constraint Propagation (AC-3)
bool solveSudokuAC3Backtracking(vector<vector<int>>& board, vector<vector<set<int>>>& domains) {
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                for (int num : domains[row][col]) {
                    if (isValid(board, row, col, num)) {
                        board[row][col] = num;
                        vector<vector<set<int>>> nextDomains = domains; // Copy domains
                        // Remove values from the domains of related cells.
                        for (int i = 0; i < 9; ++i) {
                            nextDomains[row][i].erase(num);
                            nextDomains[i][col].erase(num);
                        }
                        int startRow = row - row % 3;
                        int startCol = col - col % 3;
                        for (int i = 0; i < 3; ++i) {
                            for (int j = 0; j < 3; ++j) {
                                nextDomains[startRow + i][startCol + j].erase(num);
                            }
                        }

                        if (solveSudokuAC3Backtracking(board, nextDomains)) {
                            return true;
                        }
                        else {
                            board[row][col] = 0;
                        }
                    }
                }
                return false;
            }
        }
    }
    return true;
}

bool solveSudokuAC3(vector<vector<int>>& board) {
    // Initialize domains for each cell (1-9 for empty cells, fixed value for filled cells)
    vector<vector<set<int>>> domains(9, vector<set<int>>(9));
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                for (int num = 1; num <= 9; ++num) {
                    domains[row][col].insert(num);
                }
            }
            else {
                domains[row][col].insert(board[row][col]);
            }
        }
    }

    // Function to get the neighbors of a cell (cells in the same row, column, or block)
    auto getNeighbors = [&](int row, int col) -> vector<pair<int, int>> {
        vector<pair<int, int>> neighbors;
        for (int i = 0; i < 9; ++i) {
            if (i != col) neighbors.push_back({row, i}); // Row
            if (i != row) neighbors.push_back({i, col}); // Column
        }
        int startRow = row - row % 3;
        int startCol = col - col % 3;
        for (int i = 0; i < 3; ++i) {
            for (int j = 0; j < 3; ++j) {
                if (startRow + i != row || startCol + j != col) {
                    neighbors.push_back({startRow + i, startCol + j});
                }
            }
        }
        return neighbors;
    };

    // AC-3 Algorithm
    queue<pair<int, int>> q;
    for (int i = 0; i < 9; ++i) {
        for (int j = 0; j < 9; ++j) {
            if (board[i][j] == 0) {
                q.push({i, j}); // Start with all empty cells
            }
        }
    }

    while (!q.empty()) {
        pair<int, int> current = q.front();
        q.pop();
        int row = current.first;
        int col = current.second;

        for (pair<int, int> neighbor : getNeighbors(row, col)) {
            int nRow = neighbor.first;
            int nCol = neighbor.second;
            if (board[nRow][nCol] == 0) { // Only consider neighbors that are also unassigned.
                bool removed = false;
                for (int val : domains[row][col]) {
                    bool foundMatch = false;
                    for (int nVal : domains[nRow][nCol]) {
                        if (val != nVal) { // There exists a value in neighbor's domain that is different
                            foundMatch = true;
                            break;
                        }
                    }
                    if (!foundMatch) {
                        domains[row][col].erase(val); // Remove the value from domain
                        removed = true;
                        break; // Important:  Only remove ONCE per value in current cell's domain.
                    }
                }
                if (removed) {
                    if (domains[row][col].empty()) {
                        return false; // Domain wipeout: no solution
                    }
                    q.push({row, col}); // Re-add to queue if domain is reduced
                }
            }
        }
    }

    // Backtracking search with updated domains
    return solveSudokuAC3Backtracking(board, domains);
}

// 5. Optimized Backtracking (Combining MRV and Forward Checking) -  Most Efficient
bool solveSudokuOptimizedHelper(vector<vector<int>>& board, vector<vector<set<int>>>& possibleValues) {
    int minRow = -1, minCol = -1, minRemaining = 10;

    // Find cell with Minimum Remaining Values (MRV)
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                int count = possibleValues[row][col].size();
                if (count < minRemaining) {
                    minRemaining = count;
                    minRow = row;
                    minCol = col;
                }
            }
        }
    }

    if (minRow == -1 && minCol == -1) {
        return true; // Solved
    }

    // Iterate through possible values of the cell with MRV
    for (int num : possibleValues[minRow][minCol]) {
        if (isValid(board, minRow, minCol, num)) {
            board[minRow][minCol] = num;
            vector<vector<set<int>>> nextPossibleValues = possibleValues;

            // Forward checking: update possible values for related cells
            for (int i = 0; i < 9; ++i) {
                nextPossibleValues[minRow][i].erase(num);
                nextPossibleValues[i][minCol].erase(num);
            }
            int startRow = minRow - minRow % 3;
            int startCol = minCol - minCol % 3;
            for (int i = 0; i < 3; ++i) {
                for (int j = 0; j < 3; ++j) {
                    nextPossibleValues[startRow + i][startCol + j].erase(num);
                }
            }
            if (solveSudokuOptimizedHelper(board, nextPossibleValues)) {
                return true;
            }
            else {
                board[minRow][minCol] = 0;
            }
        }
    }
    return false;
}

bool solveSudokuOptimized(vector<vector<int>>& board) {
    vector<vector<set<int>>> possibleValues(9, vector<set<int>>(9));

    // Initialize possible values
    for (int row = 0; row < 9; ++row) {
        for (int col = 0; col < 9; ++col) {
            if (board[row][col] == 0) {
                for (int num = 1; num <= 9; ++num) {
                    if (isValid(board, row, col, num)) {
                        possibleValues[row][col].insert(num);
                    }
                }
            }
        }
    }
    return solveSudokuOptimizedHelper(board, possibleValues);
}

int main() {
    // Example Sudoku board (0 represents empty cells)
    vector<vector<int>> board = {
        {5, 3, 0, 0, 7, 0, 0, 0, 0},
        {6, 0, 0, 1, 9, 5, 0, 0, 0},
        {0, 9, 8, 0, 0, 0, 0, 6, 0},
        {8, 0, 0, 0, 6, 0, 0, 0, 3},
        {4, 0, 0, 8, 0, 3, 0, 0, 1},
        {7, 0, 0, 0, 2, 0, 0, 0, 6},
        {0, 6, 0, 0, 0, 0, 2, 8, 0},
        {0, 0, 0, 4, 1, 9, 0, 0, 5},
        {0, 0, 0, 0, 8, 0, 0, 7, 9}
    };

    cout << "Original Sudoku Board:" << endl;
    printBoard(board);

    // Create a copy for each algorithm
    vector<vector<int>> board1 = board;
    vector<vector<int>> board2 = board;
    vector<vector<int>> board3 = board;
    vector<vector<int>> board4 = board;
    vector<vector<int>> board5 = board;

    // 1. Backtracking
    cout << "1. Backtracking:" << endl;
    auto start1 = chrono::high_resolution_clock::now();
    if (solveSudokuBacktracking(board1)) {
        auto stop1 = chrono::high_resolution_clock::now();
        auto duration1 = chrono::duration_cast<chrono::microseconds>(stop1 - start1);
        cout << "Solved:" << endl;
        printBoard(board1);
        cout << "Time taken: " << duration1.count() << " microseconds" << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 2. Backtracking with Forward Checking
    cout << "2. Backtracking with Forward Checking:" << endl;
    auto start2 = chrono::high_resolution_clock::now();
    if (solveSudokuForwardChecking(board2)) {
        auto stop2 = chrono::high_resolution_clock::now();
        auto duration2 = chrono::duration_cast<chrono::microseconds>(stop2 - start2);
        cout << "Solved:" << endl;
        printBoard(board2);
        cout << "Time taken: " << duration2.count() << " microseconds" << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 3. Backtracking with MRV
    cout << "3. Backtracking with MRV:" << endl;
    auto start3 = chrono::high_resolution_clock::now();
    if (solveSudokuMRV(board3)) {
        auto stop3 = chrono::high_resolution_clock::now();
        auto duration3 = chrono::duration_cast<chrono::microseconds>(stop3 - start3);
        cout << "Solved:" << endl;
        printBoard(board3);
        cout << "Time taken: " << duration3.count() << " microseconds" << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 4. Backtracking with AC-3
    cout << "4. Backtracking with AC-3:" << endl;
    auto start4 = chrono::high_resolution_clock::now();
    if (solveSudokuAC3(board4)) {
        auto stop4 = chrono::high_resolution_clock::now();
        auto duration4 = chrono::duration_cast<chrono::microseconds>(stop4 - start4);
        cout << "Solved:" << endl;
        printBoard(board4);
        cout << "Time taken: " << duration4.count() << " microseconds" << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 5. Optimized Backtracking (MRV + Forward Checking)
    cout << "5. Optimized Backtracking (MRV + Forward Checking):" << endl;
    auto start5 = chrono::high_resolution_clock::now();
    if (solveSudokuOptimized(board5)) {
        auto stop5 = chrono::high_resolution_clock::now();
        auto duration5 = chrono::duration_cast<chrono::microseconds>(stop5 - start5);
        cout << "Solved:" << endl;
        printBoard(board5);
        cout << "Time taken: " << duration5.count() << " microseconds" << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    return 0;
}

